src/build/cargo.rs

```diff
@@ -296,38 +295,42 @@ impl Executor for RlsExecutor {
             }
         }
 
-        let rustc_exe = env::var("RUSTC").unwrap_or(env::args().next().unwrap());
-        let mut cmd = Command::new(&rustc_exe);
+        // Prepare our own call to `rustc` as follows:
+        // 1. Use $RUSTC wrapper if specified, otherwise use RLS executable
+        //    as an rustc shim (needed to distribute via the stable channel)
+        // 2. For non-primary packages or build scripts, execute the call
+        // 3. Otherwise, we'll want to use the compilation to drive the analysis:
+        //    i.  Modify arguments to account for the RLS settings (e.g.
+        //        compiling under cfg(test) mode or passing a custom sysroot)
+        //    ii. Execute the call and store the final args/envs to be used for
+        //        later in-process execution of the compiler
+        let mut cmd = cargo_cmd.clone();
+        let rls_executable = env::args().next().unwrap();
+        cmd.program(env::var("RUSTC").unwrap_or(rls_executable));
         cmd.env(::RUSTC_SHIM_ENV_VAR_NAME, "1");
 
         // We only want to intercept rustc call targeting current crate to cache
         // args/envs generated by cargo so we can run only rustc later ourselves
         // Currently we don't cache nor modify build script args
-        let is_build_script = crate_name == "build_script_build";
+        let is_build_script = *target.kind() == TargetKind::CustomBuild;
         if !self.is_primary_crate(id) || is_build_script {
-            let mut cmd = Command::new(&env::var("RUSTC").unwrap_or("rustc".to_owned()));
```

Comment:
@nrc we should still run shim for non-blacklisted crates that are deps/build scripts, right?

Reply:
Yes, any crate we want save-analysis from we must use the shim

Reply:
We should, but not for black-listed crates. Note that currently we only use cmd for blacklisted crates, so this is not wrong atm (but it should really be inside the next `if`).